Feature/add support for databricks and bigquery

agnes-kiss opened the pull request:
## Description & motivation

- [ ] The main goal with this PR is adding Databricks and BigQuery support.
- [ ] Also adding an Integration testing suite to check if the basic setup works. Also added unit tests around the path transformations. 
- [ ] Introduced a basic safety check to block users from running the package if the page_view source is not up-to-date. 

## Checklist
- [x] I have verified that these changes work locally
- [ ] I have updated the README.md (if applicable)
- [x] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have raised a [documentation](https://github.com/snowplow/documentation) PR if applicable (Link here if required)

agnes-kiss committed:
Add support for Databricks and BigQuery

rlh1994 commented on integration_tests/.gitignore:
```diff
@@ -0,0 +1,4 @@
+
+target/
+dbt_modules/
+logs/
```
it should be fine but can you add the other ones to this as well please

rlh1994 commented on integration_tests/ci/profiles.yml:
```diff
@@ -0,0 +1,78 @@
+      pass: "{{ env_var('REDSHIFT_TEST_PASS') }}"
+      dbname: "{{ env_var('REDSHIFT_TEST_DBNAME') }}"
+      port: "{{ env_var('REDSHIFT_TEST_PORT') | as_number }}"
+      schema: "gh_sp_web_dbt_{{ env_var('SCHEMA_SUFFIX') }}"
```
Schema name needs changing from `web` to `fract`

rlh1994 commented on integration_tests/README.md:
```diff
@@ -0,0 +1,18 @@
+
+The `./scripts` directory contains two scripts:
+
+- `integration_tests.sh`: This tests the standard modules of the snowplow-fractribution package. It runs the Snowplow web package to generate the derived.snowplow_we_page_views as a source then runs the package with the basic setup.
```
`web_page_views`

agnes-kiss replied:
Thanks, also added bigquery as I noticed it is available now

rlh1994 replied:
The typo is still here :D Also, the directory only has 1 script.

rlh1994 commented on integration_tests/models/source/bigquery/snowplow_fractribution_events_stg.sql:
```diff
@@ -0,0 +1,146 @@
+  se_value,
+  tr_orderid,
+  tr_affiliation,
+  case when left(cast(tr_total as string),5) = '20.42' then 20.42 else tr_total end tr_total,
```
Can you fix this by setting it as a decimal with precision in the seed instead of this?

rlh1994 commented on integration_tests/models/source/databricks/snowplow_fractribution_events_stg.sql:
```diff
@@ -0,0 +1,143 @@
+  se_value,
+  tr_orderid,
+  tr_affiliation,
+  case when left(tr_total,5) = 20.42 then 20.42 else tr_total end tr_total,
```
seed as decimal?

rlh1994 commented on integration_tests/models/unit_tests/test_path_transformations_expected.sql:
```diff
@@ -0,0 +1,88 @@
+    'Example, Video, Direct' as exposure_path,
+    'Example, Video, Direct' as first_path,
+    'Example, Video' as remove_if_not_all,
+    'Example, Video' as remove_if_last_and_not_all
```
Love this! Love it so much I want more - could you add an example where every one is different?

rlh1994 commented on integration_tests/models/unit_tests/test_path_transformations_actual.sql:
```diff
@@ -0,0 +1,55 @@
+
+    , arrays as (
+      select
+        transformed_path as raw_array,
+        {{ snowplow_fractribution.trim_long_path('transformed_path', 1) }} as trim_long_path,
+        {{ snowplow_fractribution.path_transformation('unique_path') }} as unique_path,
```
What's the reason behind `trim_long_path` getting it's own macro but the rest being part of `path_transformation`?

agnes-kiss replied:
@zloff I was meant to ask this already but forgot, is there any reason behind applying the trim_long_path() on the path only but not on the transformed_path? See [original code](https://github.com/snowplow/dbt-snowplow-fractribution/blob/96acd8a1124266ee3ad73169d08dd6a46fd668ac/models/snowflake/snowplow_fractribution_paths_to_conversion.sql#L11). Because of this I treat them separately in the model. Thanks.

zloff replied:
From what I understand, Mike stuck to the logic of the original sql when possible, looking at the code from Google, it seems that they did it this way: https://github.com/google/fractribution/blob/72ad8a4a851c75152c700283fbc4f58450b083af/py/templates/paths_to_conversion.sql#L25 
I couldn't say what the reasoning is though

rlh1994 commented on macros/path_transformations/create_udfs.sql:
```diff
@@ -0,0 +1,330 @@
+  {% set trim_long_path %}
+  -- Returns the last path_lookback_steps channels in the path if path_lookback_steps > 0,
+  -- or the full path otherwise.
+  CREATE FUNCTION IF NOT EXISTS {{schema}}.trim_long_path(path ARRAY, path_lookback_steps DOUBLE)
```
Why is `path_lookback_steps` a double here but an int for bigquery?

agnes-kiss replied:
@zloff Any reason here this should be double and not just int? I guess it is fine to have them as ints everywhere, right?

zloff replied:
Yeah no idea why it's a double, feel free to change

agnes-kiss replied:
That's why :D we all learned something today:[ Integers and Doubles](https://docs.snowflake.com/en/developer-guide/udf/javascript/udf-javascript-introduction.html#integers-and-doubles)

JavaScript has no integer type; all numbers are represented as doubles. JavaScript UDFs do not accept or return integer values except through type conversion (i.e. you can pass an integer to a JavaScript UDF that accepts a double).

rlh1994 replied:
Wait so why does it work for bigquery? Some internal translation?

agnes-kiss replied:
Must be, if I recall correctly I got this solution from the error message as a recommended workaround.

rlh1994 reviewed (changes requested):
Okay, reviewed the intergation tests sub-project so far, few tweaks/questions. Will get to main project next.

rlh1994 commented on macros/path_transformations/trim_long_path.sql:
```diff
@@ -0,0 +1,26 @@
+
+  {% if lookback_steps > 0 %}
+
+    slice(reverse({{ array_column }}), 1, (greatest(1, (cast({{ lookback_steps }} as int)))))
```
I haven't used it, but is there a reason to not do `slice({{ array_column }}, -lookback_steps, lookback_steps)` and set the default of the var in the call to 1? 

Does either the current way or the suggested way fail if the lookback is greater than the length of the array?

agnes-kiss replied:
You are right, I did not actually add the length limitation in the end. Will do that and the more elegant solution.

rlh1994 commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,47 @@
+    case when array_distinct(transformed_path) != array('{{ transform_param }}')
+    then slice(transformed_path, 1, array_size(transformed_path) - array_size(
+    filter(transformed_path, ({{ source_table }}, i) -> array_except(slice(reverse(transformed_path), 1, i), array('{{ transform_param }}'))==array()) ) + 1)
+    else transformed_path end
```
I'm sure this works, but I can't really follow what it does. Could we either add comments to explain each part please :)

rlh1994 commented on integration_tests/models/unit_tests/test_path_transformations_expected.sql:
```diff
@@ -0,0 +1,100 @@
+    '' as exposure_path,
+    '' as first_path,
+    '' as remove_if_not_all,
+    '' as remove_if_last_and_not_all
```
Is this reasonable behaviour here? They technically have a zero length path but we give thge frequency of a (1) to nothing?

agnes-kiss replied:
It is up to the user how they want to configure their [channel classification](https://github.com/snowplow/dbt-snowplow-fractribution/blob/main/macros/channel_classification.sql), in theory they would not add '' to either of their case statements but they could for specific reasons, in which case they might want to count those "paths" that are empty.

rlh1994 replied:
You've talked yourself into another test case, what about when the array contains (but is not made up entirely of) empty strings?

agnessnowplow replied:
I will try but it is not exactly scientific as some has double purpose and only valid for some of the transformations

rlh1994 commented on integration_tests/models/unit_tests/test_path_transformations_expected.sql:
```diff
@@ -0,0 +1,100 @@
+    'Direct' as exposure_path,
+    'Direct' as first_path,
+    'Direct, Direct' as remove_if_not_all,
+    'Direct, Direct' as remove_if_last_and_not_all
```
Sorry, another thing on these, can you just add a comment above eahc one where it described _what_ that test is (e.g. all the same, all different, repeat in middle, etc) to make it easier to see what has/has not been tested.

rlh1994 commented on integration_tests/packages.yml:
```diff
@@ -0,0 +1,6 @@
+  - git: "https://github.com/snowplow/dbt-snowplow-utils.git"
+    revision: "fix/cross_db_macros"
+  - git: "https://github.com/snowplow/dbt-snowplow-web.git"
+    revision: "feature/consent-adapter-support"
```
You'll know anyway but just a reminder to change this incase we lose it in all the fix commits

rlh1994 commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,47 @@
+    then array_remove(transformed_path, '{{ transform_param }}')
+    else transformed_path end
+
+  {% elif transformation_type == 'remove_if_last_and_not_all' %}
+    case when array_distinct(transformed_path) != array('{{ transform_param }}')
+    then slice(transformed_path, 1, array_size(transformed_path) - array_size(
+    filter(transformed_path, (x, i) -> array_except(slice(reverse(transformed_path), 1, i), array('{{ transform_param }}'))==array()) ) + 1)
+    else transformed_path end
```
Could you add some comments around this one please? I can follow the rest but this is giving me a headache :)

agnes-kiss replied:
It's Mike's clever sql workaround but I added a high level summary. As I understand it checks for any occurrences of the parameter at the end of the list and in order to remove them with the first slice it needs to calculate how many of these to remove, which it gets by the filter bits when he reverses the path to slice it and then by counting it with the array_size. Not sure I can explain it in plain English well, hence the high level summary you'll see.

rlh1994 replied:
@miike could you possibly de-mystify how exactly this works for us please? It's clear up until the last argument of the first `slice` but that part is the work of a ~madman~ genius

miike replied:
Yeah this is not super clear (but does avoid us having to use UDFs) and a couple of works around for functions that don't exist in Databricks.

The main aim of this function is to remove any trailing instances of the input parameter (usually Direct) where there is one or more instances and they are not the only value in the array (the first part of the case checks for this).

The last argument to the first slice is designed to give us the offset of how far we need to slice in the array to remove those trailing references it does this by:
- iteratively slicing (using the i param in the filter) the array and determining if it only contains direct references (using array_except) which we then return an element for only if they are equivalent. We do this in reverse as we want to count the trailing references.

From an iteration / slice perspective this looks like (for the following example): 
```
["Example", "Another", "Direct", "Direct"]
filter(y, (x, i) -> array_except(slice(reverse(y), 1, i), array('Direct'))==array())

Slice 1 (i=1): Direct.
array_except yields [] as our array only contains 'Direct' references, comparison yields True
Slice 2 (i=2): Direct, Direct
array_except yields [], comparison yields True
Slice 3 (i=3): Direct, Direct, Another
array_except yields [Another], comparison yields False (element does not become part of the array)
Slice 4 (i=4): Direct, Direct, Another, Example
array_except yields [Another, Example], comparison yields False (element does not become part of the array)

At this point we can now count the size of this array - which gives us an index (from the back of the array) as to how many elements we can chop off - so to convert this to a an actual slice (as negative slicing sort of works in DB) we do:
```array_size(original) - array_size(direct_size) + 1
4 - 2 + 1 = 3
```
and then slice (using 1-indexed array) from the first element to the starting point of the repeated directs and that gives us our result!

rlh1994 commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,47 @@
+    else transformed_path end
+
+  {% else %}
+    {%- do exceptions.raise_compiler_error("Snowplow Warning: the path transform - '"+transformation_type+"' - is not yet supported for Databricks. Please choose from the following: exposure_path, first_path, remove_if_not_all, unique_path") %}
+
```
Needlessly picky, should this say `Snowplow Error:` instead?

agnes-kiss replied:
Not picky, I also thought of this but never bothered me enough to actually change it. But will do, better be clean.

rlh1994 commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,47 @@
+{% macro path_transformation(transformation_type, transform_param) %}
+  {{ return(adapter.dispatch('path_transformation', 'snowplow_fractribution')(transformation_type, transform_param)) }}
+{% endmacro %}
+
+-- only used for integration tests
+{% macro default__path_transformation(transformation_type, transform_param) %}
+
```
used in general now isn't it?

agnes-kiss replied:
I don't think so. The transform_path macro will use the same code as this macro for Snowflake/Bigquery. It is to avoid the issue we faced when trying to generate a macro call with a parameter.

rlh1994 commented on macros/schema.yml:
```diff
@@ -12,6 +12,3 @@ macros:
 
   - name: conversion_value
     description: A macro that specifies either a single column or a calculated value that represents the value associated with the conversion
```
Could you add all the macros, including path transform ones, into here including args and other parts from the style guide please

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,125 @@
+    {% endif %}
+
+  from arrays
+
```
I know we're only using the macro in this package, so it's not for end users to customise, but in case we edit the CTE name in the main model could we make this CTE name an argument to the function? It might just save us some debugging down the line if we forget this is hardcoded

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,125 @@
+{% macro databricks__transform_paths(model_type) %}
+
+  {% set total_transformations = var('path_transforms').items()|length %}
+  {% set loop_count = namespace(value=1) %}
```
the `namespace` call is how you set a global variable right? could you comment that here as it's not something that is often used

agnes-kiss replied:
Yes, it is a workaround for the loop to work if defined outside.

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,125 @@
+  {% set total_transformations = var('path_transforms').items()|length %}
+  {% set loop_count = namespace(value=1) %}
+
+  {% for path_transform_name, transform_param in var('path_transforms').items() %}
+
+    {%- if loop_count.value == 1 %}
```
Does this loop need to be inside a `{% if total_transformations > 0 %}` check? Or does the loop just not do anything if the dictionary is empty?

agnes-kiss replied:
No (tested it), it will skip the path transform loop altogether and go do the trim long path one only: 
<img width="695" alt="image" src="https://user-images.githubusercontent.com/95634439/216138538-46e7e288-e5da-4a1d-a813-ee1c1c1623fa.png">

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,125 @@
+  {% set total_transformations = var('path_transforms').items()|length %}
+  {% set loop_count = namespace(value=1) %}
+
+  {% for path_transform_name, transform_param in var('path_transforms').items() %}
+
+    {%- if loop_count.value == 1 %}
```
Is this loop a different order to the `default__` one? in default you reverse the list so that you call the last one first, that gets used by the second to last etc. In this one you are going front to back I think?

agnes-kiss replied:
Yes, but I believe here it is not needed as each transformation is done in a new cte whereas in the other you are nesting functions and their parameters, one of which has to be reversed. Which I just noticed was wrong for snowflake so fixed that.

rlh1994 replied:
Could this not lead to different results though? E.g. if you do a unique then do the frequency you'd get different than the other way round. The CTEs are still just nested transformations, so we should match the order between both methods?

agnes-kiss replied:
Agreed, I will reverse the reverse.

agnes-kiss replied:
And will also specify the order of processing in the comments as it is not clear for the user either.

rlh1994 replied:
I'm inclined to say keep the reverse (more complexity for us, add it to the databricks one) as it means it's then applied in the same order it's written right (more common for the user)? But I agree add to the variable description as well :)

agnes-kiss replied:
I meant it that way actually. Reverse the reverse as in change from this:     
 ```
 {% for path_transform_name, _ in var('path_transforms').items()|reverse %}
        {% if path_transform_name not in allowed_path_transforms %}
          {%- do exceptions.raise_compiler_error("Snowplow Error: the path transform - '"+path_transform_name+"' - is not supported. Please refer to the Snowplow docs on tagging. Please use one of the following: exposure_path, first_path, frequency_path, remove_if_last_and_not_all, remove_if_not_all, unique_path") %}
        {% endif %}
        {{target.schema}}.{{path_transform_name}}(
      {% endfor %}

      transformed_path

      {% for _, transform_param in var('path_transforms').items() %}
        {% if transform_param %}, '{{transform_param}}' {% endif %}
        )
      {% endfor %}
```
To the the other way round to keep the order the same as for snowflake processing left to right.

agnes-kiss replied:
I mean change to this for snowflake and do nothing for databricks, I think we clarified it separately but will add comments as it is confusing, should have added it already

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,125 @@
+         )
+        {% endif %}
+        {% set loop_count.value = loop_count.value + 1 %}
+        {% set previous_cte = loop_count.value-1 %}
```
Smart, in genreal can we get more comments across this and other macro files for dumb people (me) please :)

rlh1994 commented on macros/get_lookback_date_limits.sql:
```diff
@@ -0,0 +1,68 @@
+  -- check if web data is up-to-date
+
+  {% set query %}
+    with base as (
+      select max(start_tstamp) max_start_tstamp
+      from {{ var('page_views_source') }}
+    )
+
+    select
+      case when max_start_tstamp  < '{{ var('conversion_window_end_date') }}'
+      then True else False end as is_over_limit
+    from base
```
does `select max(start_tstamp) < '{{ var('conversion_window_end_date') }}' as is_over_limit from {{ var('page_views_source') }}` work?

agnes-kiss replied:
Definitely more elegant, will change the fast-script code.

rlh1994 commented on macros/get_lookback_date_limits.sql:
```diff
@@ -0,0 +1,68 @@
+    {% endif %}
+
+  {% else %}
+  {% endif %}
```
Could simplify this by just putting the set in the if blocks and then getting the result and returning it outside, just reduced some duplication. You also have an eronious `else` block

rlh1994 commented on models/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -0,0 +1,53 @@
+    c.customer_id,
+    c.conversion_tstamp,
+    revenue,
+    {{ snowplow_utils.get_string_agg('channel', 's', separator=' > ', sort_numeric=false, order_by_column='visit_start_tstamp', partition_by_columns='c.customer_id', order_by_column_prefix='s') }} as path,
+    {{ snowplow_utils.get_string_agg('channel', 's', separator=' > ', sort_numeric=false, order_by_column='visit_start_tstamp', partition_by_columns='c.customer_id', order_by_column_prefix='s') }} as transformed_path
```
This is the same thing, could you just do path and use that in the next part?

rlh1994 commented on models/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -0,0 +1,53 @@
+
+  from {{ ref('snowplow_fractribution_conversions_by_customer_id') }} c
+
+  left join {{ ref('snowplow_fractribution_sessions_by_customer_id') }} s
+  on c.customer_id = s.customer_id
+    and {{ datediff('s.visit_start_tstamp', 'c.conversion_tstamp', 'day') }}  >= 0
+    and {{ datediff('s.visit_start_tstamp', 'c.conversion_tstamp', 'day') }} <= {{ var('path_lookback_days') }}
```
Given we need the match for the order by, should this explictly be an inner join?

agnes-kiss replied:
All the three order by columns are coming from the main table so not sure if it is essential but I think the inner join should also be fine as we expect to have page_views information leading to conversion in the first place. We have an upper boundary check earlier on to see if the processed pageview data covers the analysis period.

rlh1994 commented on models/snowplow_fractribution_paths_to_non_conversion.sql:
```diff
@@ -0,0 +1,72 @@
+-- Requires macro trim_long_path
+
+
+with conversions as (
+
+  select distinct customer_id
+
+  from {{ ref('snowplow_fractribution_conversions_by_customer_id') }}
+)
+
+, non_conversions as (
+
+  select
+    s.customer_id,
+    max(s.visit_start_tstamp) as non_conversion_tstamp
+
+  from {{ ref('snowplow_fractribution_sessions_by_customer_id') }} s
+
+  left join conversions c
+    on s.customer_id = c.customer_id
+
+  where c.customer_id is null
+
+  group by 1
+)
```
Swap this out for a `not exists` clause instead?

rlh1994 commented on models/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -0,0 +1,53 @@
+    and {{ datediff('s.visit_start_tstamp', 'c.conversion_tstamp', 'day') }}  >= 0
+    and {{ datediff('s.visit_start_tstamp', 'c.conversion_tstamp', 'day') }} <= {{ var('path_lookback_days') }}
+
+{% if target.type in ['snowflake', 'bigquery'] -%}
+  group by 1,2,3
+{%- endif %}
```
why don't we group for databricks?

agnes-kiss replied:
Because the workaround solution for the utils' databricks__get_string_agg (to be able to sort based on another column) involves using a non-aggregate window function, whereas for the rest we have aggregate functions like string_agg for bigquery.

rlh1994 replied:
Does that not mean we need to do a `distinct` for databricks? I also absolutely hate that we have different return types for the same function on the same warehouse, let alone across warehouses...

rlh1994 replied:
I guess also for future-proofing it should be `if target.type not in ['databricks', 'spark']` ?

agnes-kiss replied:
Yes, that's why I added 'distinct', it is for databricks only. Definitely not ideal but that's the only solution for sorting based on another column.

rlh1994 commented on models/snowplow_fractribution_paths_to_non_conversion.sql:
```diff
@@ -0,0 +1,72 @@
+
+  select distinct
+    n.customer_id,
+    {{ snowplow_utils.get_string_agg('channel', 's', separator=' > ', order_by_column='visit_start_tstamp', sort_numeric=false, partition_by_columns='n.customer_id', order_by_column_prefix='s') }} as path,
+    {{ snowplow_utils.get_string_agg('channel', 's', separator=' > ', order_by_column='visit_start_tstamp', sort_numeric=false, partition_by_columns='n.customer_id', order_by_column_prefix='s') }} as transformed_path
+
+
+  from non_conversions n
+
+  left join {{ ref('snowplow_fractribution_sessions_by_customer_id') }} s
+  on n.customer_id = s.customer_id
+    and {{ datediff('s.visit_start_tstamp', 'n.non_conversion_tstamp', 'day') }}  >= 0
```
Same as conversions, only need to do column once and inner join?

rlh1994 commented on models/snowplow_fractribution_sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,53 @@
+
+    and date(derived_tstamp) <= '{{ get_lookback_date_limits('max') }}'
+
+    and page_urlhost in (
+        {%- for urlhost in var('conversion_hosts') %}
+            '{{ urlhost }}'
+            {%- if not loop.last %},{% endif %}
+        {%- endfor %}
+    )
```
Could use `snowplow_utils.print_list()` here?

agnes-kiss replied:
Was not aware it existed but for sure.

rlh1994 commented on models/snowflake/snowplow_fractribution_sessions_by_customer_id.sql:
```diff
@@ -1,70 +0,0 @@
-        AND page_view_in_session_index = 1 -- takes the first page view in the session
-    {% endif %}
-    
-    {% if var('channels_to_exclude') %}
-        -- Filters out any unwanted channels
-        AND channel NOT IN (
-            {%- for unwanted_channel in var('channels_to_exclude') %}
-                '{{ unwanted_channel }}'
-                {%- if not loop.last %},{% endif %}
-            {%- endfor %}
-        )
-    {% endif %}
```
Did you lose this somewhere? Can't see it in the new one?

agnes-kiss replied:
I haven't rebased it yet, it was added as part of the new release on another branch.

agnes-kiss replied:
will change this to utils print_list as well

rlh1994 commented on models/snowplow_fractribution_sessions_by_customer_id.sql:
```diff
@@ -0,0 +1,53 @@
+  on page_views.domain_userid = user_mapping.domain_userid
+{% endif %}
+
+where date(derived_tstamp) >= '{{ get_lookback_date_limits('min') }}'
+
+    and date(derived_tstamp) <= '{{ get_lookback_date_limits('max') }}'
```
syntax highlighting upsets me, can you put the internal ones in double quotes please

rlh1994 commented on models/snowplow_fractribution_conversions_by_customer_id.sql:
```diff
@@ -0,0 +1,32 @@
+    on events.domain_userid = user_mapping.domain_userid
+{% endif %}
+
+where {{ conversion_clause() }}
+  and date(derived_tstamp) >= case when '{{ var('conversion_window_start_date') }}' = ''
+                                then current_date()-31
+                                else '{{ var('conversion_window_start_date') }}'
+                                end
+  and date(derived_tstamp) <= case when '{{ var('conversion_window_end_date') }}' = ''
+                                then current_date()-1
+                                else '{{ var('conversion_window_end_date') }}'
+                                end
```
Can you use the lookback window macro here or no?

agnes-kiss replied:
Yeah, should be the same here.

rlh1994 reviewed (changes requested):
I have basically no memory of what all the comments I put are, but I feel like they were mostly comments/optimisations/syntax stuff rather than actual problems!

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,126 @@
+  {% for path_transform_name, transform_param in var('path_transforms').items() %}
+
+    {%- if loop_count.value == 1 %}
+      {% set previous_cte = 'arrays' %}
```
Need to add the new `source_cte` arg here as well I think? (even though I do think it never hits this on the first itteration so maybe it's not needed at all)

agnes-kiss replied:
I noticed this already while fixing the above. It should hit this otherwise it would fail as these are mid script ctes.

rlh1994 commented on macros/schema.yml:
```diff
@@ -3,12 +3,44 @@ version: 2
+    arguments:
+      - name: transformation_type
+        type: string
+        description: A type of transformation that needs to be executed E.g. 'unique_path'
```
Should we list the specific options (for databricks) here? (as well as it's already in the main docs)

rlh1994 replied:
Also maybe we could put a link in here or the macro description to https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/dbt-fractribution-data-model/#path-transform-options just to save duplicating words, I find having the examples on hand infinitely valuable (side note that page needs the two `remove` options added)

zloff commented on dbt_project.yml:
```diff
@@ -26,9 +26,10 @@ vars:
   conversion_window_start_date: '' # conversion window start date
   conversion_window_end_date: '' # conversion window end date is dynamically set to current_date()-1 in the code
   conversion_hosts: ['poplindata.com'] # url_hosts to consider
```
did we want to change this to 'a.com' like what is in the sample data, or is that being done somewhere else?

agnes-kiss replied:
First I was thinking of changing it in the accelerator instructions but good point, we have the chance to change it in the source here for future users.

zloff commented on models/snowplow_fractribution_conversions_by_customer_id.sql:
```diff
@@ -0,0 +1,32 @@
+{% endif %}
+
+where {{ conversion_clause() }}
+  and date(derived_tstamp) >= case when '{{ var('conversion_window_start_date') }}' = ''
```
Maybe we should also change the quotes here for syntax highlighting/Ryan's sanity purposes :P

miike replied:
On this - do we have a preferred linter / standard for these things going forward?

rlh1994 replied:
Not at the moment, we're probably going to start using sqlfluff when 2.0 is released, and work to build up our style guide over time enforced by that

zloff commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -0,0 +1,126 @@
+        {% endif %}
+        {% set loop_count.value = loop_count.value + 1 %}
+        {% set previous_cte = loop_count.value-1 %}
+
```
Would it be clearer if we swapped lines 102 and 103? So:
```
{% set previous_cte = loop_count.value %}
{% set loop_count.value = loop_count.value + 1 %}
```
Otherwise we increment just to deincrement (_apparently deincrement isn't a word, but i'm sticking with it_) in the next line

agnes-kiss replied:
Yeah, definitely. I think decrement is the word here.  :)

zloff commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,48 @@
+
+  {% else %}
+    {%- do exceptions.raise_compiler_error("Snowplow Error: the path transform - '"+transformation_type+"' - is not yet supported for Databricks. Please choose from the following: exposure_path, first_path, remove_if_not_all, unique_path") %}
+
```
i may be misunderstanding, but the other transformation types are working now, right? (frequency, removeiflastandnotall). Should we add them to this list of possibilities?

zloff commented on macros/get_lookback_date_limits.sql:
```diff
@@ -0,0 +1,68 @@
+      case when max_start_tstamp  < '{{ var('conversion_window_end_date') }}'
+      then True else False end as is_over_limit
+    from base
+  {% endset %}
```
Do we need to consider the other possibility, that page_views could be run on the past month of data, for example, but then the user tries to run frac on the past 2 months, etc? this check would pass in that situation, from what i understand? I imagine that situation is much less likely though than what we're already checking for here.

agnes-kiss replied:
Good point, I forgot that in the end I decided not to use the manifest where we only have the upper boundary defined. will add the lower limit check as well.

rlh1994 commented on macros/schema.yml:
```diff
@@ -2,16 +2,48 @@ version: 2
+          More details here https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/dbt-fractribution-data-model/#path-transform-options
+      - name: transform_param
+        type: string
+        description: The parameter value that the path transormation needs to execute, which is optional
```
```suggestion
description: (Optional) The parameter value that the path transormation needs to execute,. Default none
```

rlh1994 commented on macros/schema.yml:
```diff
@@ -2,16 +2,48 @@ version: 2
+        description: The array column to be transformed
+      - name: lookback_steps
+        type: integer
+        description: Defaulted to be taken from the path_lookback_steps, the number of path to leave starting from the end
```
```suggestion
description: (Optional) The number of path to leave starting from the end. Default `var(path_lookback_steps)`
```

rlh1994 commented on macros/schema.yml:
```diff
@@ -2,16 +2,48 @@ version: 2
+    arguments:
+      - name: transformation_type
+        type: string
+        description: >
+          A type of transformation that needs to be executed E.g. 'unique_path'.
+          Needs to be one of the following: exposure_path, first_path, remove_if_not_all, unique_path
+          More details here https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/dbt-fractribution-data-model/#path-transform-options
```
We offer 6 now don't we? (also need to update in the docs in this case, although `unique` probably doesn't need to go in the actual docs).

agnes-kiss replied:
The next item on my to-do is to add the docs, noticed we were missing some. Thanks for spotting this, of course I copied. this from the only wrong line that has been changed since. will add them here, too.

rlh1994 commented on macros/path_transformations/path_transformation.sql:
```diff
@@ -0,0 +1,50 @@
+    then slice(transformed_path, 1, array_size(transformed_path) - array_size(
+    filter(transformed_path, (x, i) -> array_except(slice(reverse(transformed_path), 1, i), array('{{ transform_param }}'))==array()) ) + 1)
+    else transformed_path end
+
```
```suggestion
  {% elif transformation_type == 'remove_if_last_and_not_all' %}
    /* remove the matching path(s) from the tail unless it removes everything (obtaining the upper boundary of the
    slicing to do this is done by slicing the array and determining if it only contains the desired references which
    it then returns an element for only if they are equivalent.) 
    Example:
        ["Example", "Another", "Direct", "Direct"]
        filter(y, (x, i) -> array_except(slice(reverse(y), 1, i), array('Direct'))==array())
        
        Slice 1 (i=1): Direct.
        array_except yields [] as our array only contains 'Direct' references, comparison yields True
        Slice 2 (i=2): Direct, Direct
        array_except yields [], comparison yields True
        Slice 3 (i=3): Direct, Direct, Another
        array_except yields [Another], comparison yields False (element does not become part of the array)
        Slice 4 (i=4): Direct, Direct, Another, Example
        array_except yields [Another, Example], comparison yields False (element does not become part of the array)
        
        At this point we can now count the size of this array - which gives us an index (from the back of the array) as to how many elements we can chop off - so to convert this to a an actual slice (as negative slicing sort of works in DB) we do:
        array_size(original) - array_size(direct_size) + 1
        4 - 2 + 1 = 3   
    */
    case when array_distinct(transformed_path) != array('{{ transform_param }}')
    then slice(transformed_path, 1, array_size(transformed_path) - array_size(
    filter(transformed_path, (x, i) -> array_except(slice(reverse(transformed_path), 1, i), array('{{ transform_param }}'))==array()) ) + 1)
    else transformed_path end

```

rlh1994 commented on macros/get_lookback_date_limits.sql:
```diff
@@ -0,0 +1,55 @@
+  -- check if web data is up-to-date
+
+  {% set query %}
+    select max(start_tstamp) < '{{ var('conversion_window_end_date') }}' as is_over_limit,
```
Sorry can you fix the single quotes in single quotes here please :)

rlh1994 commented on models/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -0,0 +1,53 @@
+-- Requires macro trim_long_path
+
+with string_aggs as (
+
+  select distinct
+    c.customer_id,
```
This is the tiniest performance difference if at all, but could you wrap the `distinct` in a databricks'/spark if block?

rlh1994 replied:
(Same in the non_conversions)

rlh1994 commented on macros/get_lookback_date_limits.sql:
```diff
@@ -25,44 +19,37 @@
+    {% set page_view_min = result[0][1] %}
+    {% if page_view_min == True %}
+      {%- do exceptions.raise_compiler_error("Snowplow Warning: the derived.page_view source does not cover the full fractribution analysis period. Please backfill / reprocess your web model first before proceeding with this package.") %}
+    {% endif %}
```
Could we be a little more explicit here in terms of if their min or max date isn't covered?

rlh1994 replied:
If we wanted we could also pull the min/max date and actually compare it to the variable, telling them the values so they know what they need to process to/from... up to you though that's more effort for little gain.

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+#### Usage
+
+```sql
+
+...
+{{ trim_long_path('path', var('path_lookback_steps')) }} as path,
+..
+
````
```suggestion

select
  ...
  {{ trim_long_path('path', var('path_lookback_steps')) }} as path,
  ...
from 
  ...

```

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+The transformed array column.
+
+
+#### Usage
+
+```sql
+
+```
````
Delete or fill empty usage blocks

agnes-kiss replied:
Only one block. :) I added a simple usage example.

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+#### Usage
+
+```yml
+
+on-run-start: "{{ create_udfs() }}"
+
````
```suggestion
-- dbt_project.yml
...
on-run-start: "{{ create_udfs() }}"
...

```

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+A macro call with 'min' or 'max' given as a parameter.
+
+```sql
+where date(derived_tstamp) >= '{{ get_lookback_date_limits('min') }}'
+
+    and date(derived_tstamp) <= '{{ get_lookback_date_limits('max') }}'
````
```suggestion
select
  ...
from 
  ...
where 
  date(derived_tstamp) >= '{{ get_lookback_date_limits("min") }}'
  and date(derived_tstamp) <= '{{ get_lookback_date_limits("max") }}'
  
-- returns
select
  ...
from 
  ...
where 
  date(derived_tstamp) >= '2023-01-01 13:45:03'
  and date(derived_tstamp) <= '2023-02-01 10:32:52'
```

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+
+A sql to be used to refer to the conversion value.
+
+#### Usage
+
+```sql
+
+tr_total
+
+```
````
This would be an example of the macro contents, it should be an example of the macro being called.

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+
+#### Usage
+
+```sql
+ tr_total > 0
+
+```
````
Same thing, this is the output not the usage

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+
+#### Usage
+
+```sql
+
+  -- Example (simplified) query:
+
+  select
+    channel,
+    sum(spend_usd) as spend
+  from example_spend_table
+  group by 1
+
+  -- Example table output for the user-supplied SQL:
+
+  Channel     |  Spend
+ ------------------------
+  direct      |  1050.02
+  paid_search |  10490.11
+  etc...
+
+```
````
Input vs usage

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+
+#### Usage
+
+```sql
+    case when lower(mkt_medium) in ('cpc', 'ppc') and regexp_count(lower(mkt_campaign), 'brand') > 0 then 'Paid_Search_Brand'
+         when lower(mkt_medium) in ('cpc', 'ppc') and regexp_count(lower(mkt_campaign), 'generic') > 0 then 'Paid_Search_Generic'
+         when lower(mkt_medium) in ('cpc', 'ppc') and not regexp_count(lower(mkt_campaign), 'brand|generic') > 0 then 'Paid_Search_Other'
+         when lower(mkt_medium) = 'organic' then 'Organic_Search'
+         when lower(mkt_medium) in ('display', 'cpm', 'banner') and regexp_count(lower(mkt_campaign), 'prospect') > 0 then 'Display_Prospecting'
+         when lower(mkt_medium) in ('display', 'cpm', 'banner') and regexp_count(lower(mkt_campaign), 'retargeting|re-targeting|remarketing|re-marketing') > 0 then 'Display_Retargeting'
+         when lower(mkt_medium) in ('display', 'cpm', 'banner') and not regexp_count(lower(mkt_campaign), 'prospect|retargeting|re-targeting|remarketing|re-marketing') > 0 then 'Display_Other'
+         when regexp_count(lower(mkt_campaign), 'video|youtube') > 0 or regexp_count(lower(mkt_source), 'video|youtube') > 0 then 'Video'
+         when lower(mkt_medium) = 'social' and regexp_count(lower(mkt_campaign), 'prospect') > 0 then 'Paid_Social_Prospecting'
+         when lower(mkt_medium) = 'social' and regexp_count(lower(mkt_campaign), 'retargeting|re-targeting|remarketing|re-marketing') > 0 then 'Paid_Social_Retargeting'
+         when lower(mkt_medium) = 'social' and not regexp_count(lower(mkt_campaign), 'prospect|retargeting|re-targeting|remarketing|re-marketing') > 0 then 'Paid_Social_Other'
+         when mkt_source = '(direct)' then 'Direct'
+         when lower(mkt_medium) = 'referral' then 'Referral'
+         when lower(mkt_medium) = 'email' then 'Email'
+         when lower(mkt_medium) in ('cpc', 'ppc', 'cpv', 'cpa', 'affiliates') then 'Other_Advertising'
+         else 'Unmatched_Channel'
+    end
+```
````
I think this is the definition/output again?

rlh1994 commented on docs/markdown/snowplow_fractribution_macros_docs.md:
````diff
@@ -0,0 +1,215 @@
+It is used by the transform_paths() macro for the transformation cte sql code build. It takes a transformation type as a parameter and its optional argument, if exists. The E.g.
+
+```sql
+
+{% elif path_transform_name == 'first_path' %}
+          {{ path_transformation('first_path') }} as transformed_path
+
````
```suggestion 
with base_data as (...),

{{ transform_paths('conversions', 'base_data') }}
 
select * from path_transforms
```

rlh1994 reviewed (changes requested):
Pretty much just comments on the docs and a tiny potential optimisation, otherwise it's there!

agnes-kiss commented:
I consider the dbt side of things to be ready, please let me know if there is still any open issues still. The only pending point is to potentially remove dependency on the utils' get_string_agg macro for Databricks due to the different return type, if we decide to decouple it. 

I went ahead and looked at the python scripts, added some still pending fixes for Snowflake and added as separate `main_snowplow_databricks`.py. Far from a clean code but it seems to be working. Couple of open topics worth discussing around this: 1. Unlike in case of Snowflake I do not get much feedback on the state of querying and whether the session is indeed closed. Perhaps I should add more print()s to give the user a feedback where we are? 2. For now we only have one requirements file, not sure if it's worth to keep separate versions depending on the warehouse.

rlh1994 commented on dbt_project.yml:
```diff
@@ -24,20 +24,20 @@ clean-targets:         # directories to be removed by `dbt clean`
+  snowplow__page_views_source: "{{ source('derived', 'snowplow_web_page_views') }}"
+  snowplow__conversions_source: "{{ source('atomic', 'events') }}"
 
   # snowplow_web_user_mapping_table: derived.snowplow_web_user_mapping # path (schema and table) in to the Snowplow web user mapping table (BigQuery)
```
Is this bigquery specific still?

agnes-kiss replied:
No, removed the leftover reference from snowplow_web_user_mapping_table as well

rlh1994 commented on dbt_project.yml:
```diff
@@ -24,20 +24,20 @@ clean-targets:         # directories to be removed by `dbt clean`
+  snowplow__path_lookback_steps: 0 # Limit for the number of marketing channels to look at before the conversion (0 = unlimited)
+  snowplow__path_lookback_days: 30 # Restrict to marketing channels within this many days of the conversion (recommended: 30 / 14 / 7)
 
   consider_intrasession_channels: false # false = only considers the channel at the start (first page view) of the session, true = consider multiple channels in the conversion session as well as historical channels
```
What's the reason for not prefixing this one?

agnes-kiss replied:
Nothing apart from being in a hurry. Will change it!

rlh1994 commented on macros/path_transformations/transform_paths.sql:
```diff
@@ -6,31 +6,31 @@
+  {% set allowed_snowplow__path_transforms = ['exposure_path', 'first_path', 'frequency_path', 'remove_if_last_and_not_all', 'remove_if_not_all', 'unique_path'] %}
 
-  , path_transforms as (
+  , snowplow__path_transforms as (
```
I think you might have caught a few things with a find and replace you maybe weren't meaning to? In a few other places as well, will try and tag them.

agnes-kiss replied:
Good catch, it is all due to the path_transform variable. Will sort all.

rlh1994 commented on models/snowplow_fractribution_paths_to_conversion.sql:
```diff
@@ -49,4 +49,4 @@ select
   {{ snowplow_utils.get_array_to_string('transformed_path', 'p', ' > ') }} as transformed_path
 
-from path_transforms p
+from snowplow__path_transforms p
```
find+replace error?

rlh1994 commented on models/snowplow_fractribution_paths_to_non_conversion.sql:
```diff
@@ -58,4 +58,4 @@ select
   {{ snowplow_utils.get_array_to_string('transformed_path', 'p', ' > ') }} as transformed_path
 
-from path_transforms p
+from snowplow__path_transforms p
```
find+replace error?

rlh1994 reviewed (commented):
Will check the python later

rlh1994 commented:
> 1. Unlike in case of Snowflake I do not get much feedback on the state of querying and whether the session is indeed closed. Perhaps I should add more print()s to give the user a feedback where we are? 

Can you add a `--verbose` flag maybe and only do the extra prints if they call it with that?

> 2. For now we only have one requirements file, not sure if it's worth to keep separate versions depending on the warehouse.

I think have a master one because people will just be used to a `pip install -r requirements.txt`, but also have bespoke ones to slim down if people want to only install what they need?

rlh1994 commented on utils/Dockerfile:
```diff
@@ -1,23 +1,38 @@
 
-CMD ["sh", "-c", "python main_snowplow_snowflake.py --attribution_model=$attribution_model --conversion_window_start_date=$conversion_window_start_date --conversion_window_end_date=$conversion_window_end_date"]
\ No newline at end of file
+CMD ["sh", "-c", "python main_snowplow_$warehouse.py --attribution_model=$attribution_model --conversion_window_start_date=$conversion_window_start_date --conversion_window_end_date=$conversion_window_end_date"]
```
Am I being dumb, where is this warehouse value set?

agnes-kiss replied:
I'm pasting this slack fragment from my conversations with Zoe for more context on this: 
<img width="1312" alt="image" src="https://user-images.githubusercontent.com/95634439/217896512-ceb04301-dfbc-4d20-b473-7179e7fbe01d.png">

rlh1994 replied:
Okay difference between environment variables and variables from environment files got it. I'll learn docker properly one day...

rlh1994 commented on utils/main_snowplow_bigquery.py:
```diff
@@ -0,0 +1,249 @@
+    )
+
+
+def _extract_channels(client, params: Mapping[str, Any]) -> List[str]:
+    """Returns the list of names by running extract_channels.sql.
+
+    Args:
+      client: Client.
+      params: Mapping of template parameter names to values.
+    Returns:
+      List of channel names.
+    Raises:
+      ValueError: User-formatted error if channel is not a valid Bigquery column.
+    """
```
If I'm being picky, this function doesn't use the params: argument at all

agnes-kiss replied:
I got this removed for all warehouses, none uses it in the end, must have been a leftover unused code.

rlh1994 commented on utils/main_snowplow_bigquery.py:
```diff
@@ -0,0 +1,249 @@
+
+project_id = os.environ.get('project_id')
+dataset = os.environ.get('bigquery_dataset')
+
+_OUTPUT_TABLES = ["snowplow_fractribution_path_summary_with_channels", "snowplow_fractribution_report_table", "snowplow_fractribution_channel_attribution"]
+
```
This is never used? If it's just for us then could we make it a comment?

agnes-kiss replied:
Done, you are right, not being used in the end.

rlh1994 reviewed (commented):
Just working through comparing the python to the snowflake one it all looks good, I don't have a core understanding of what it's all doing but if you and @zloff are happy with it then I'm happy with it.

Final thing I think on this PR is getting the databricks string_agg swapped to the new one, and then we just need to wait for the utils release (which is pending the consent stuff I think?)

agnes-kiss commented:
> Just working through comparing the python to the snowflake one it all looks good, I don't have a core understanding of what it's all doing but if you and @zloff are happy with it then I'm happy with it.
> 
> Final thing I think on this PR is getting the databricks string_agg swapped to the new one, and then we just need to wait for the utils release (which is pending the consent stuff I think?)

I think what's left now is to resolve questions around authentication? @zloff said she could not run the databricks python yet due to authentication issues so unless I'm wrong noone checked it actually works yet. Bigquery worked for me when I tested it. The other is checking if Docker works, which I can do once I have info on it in the docs. Thanks!

rlh1994 commented on dbt_project.yml:
```diff
@@ -16,28 +16,30 @@ test-paths: ["tests"]
+  snowplow__page_views_source: "{{ source('derived', 'snowplow_web_page_views') }}"
+  snowplow__conversions_source: "{{ source('atomic', 'events') }}"
 
-  # snowplow_web_user_mapping_table: derived.snowplow_web_user_mapping # path (schema and table) in to the Snowplow web user mapping table (BigQuery)
+  # snowplow_web_user_mapping_table: derived.snowplow_web_user_mapping # path (schema and table) to the Snowplow web user mapping table
```
Should be a double underscore

rlh1994 commented:
@zloff How hard would it be to add a `docker compose` file for the users? /is there a reason to not?

zloff commented:
> How hard would it be to add a docker compose file for the users? /is there a reason to not?

@rlh1994 It wouldn't be too hard to add one - but I think we are on the precipice of whether it would actually be useful for the user or not. Most of the time consuming part for the user is in creating the env vars, which one way or another you have to copy and paste the values anyway whether you're using compose or not. I've simplified the docker command line instructions now and just added the warehouse var to the env file, which I think makes it seem less complicated. I'll have a look at compose files today, but my instinct is that this use case may not be quite complicated enough for compose to provide any better user experience

agnes-kiss commented:
I have tested the new Docker image locally: Snowflake ran fine, BigQuery only worked if either I removed this env var altogether from the configs.env or defined the google app creds like this: `google_application_credentials=/keyfile.json` For this reason I made the change in the Dockerfile (as we ask users to refer to that for creating their confings.env in the first place) However, I encountered a network request error with Databricks when I try to run it through docker (using the same credentials as for python) `_databricks.sql.exc.RequestError: Error during request to server_ `It would be great if someone could double check if it is an issue on their side as well.

zloff commented:
> I have tested the new Docker image locally: Snowflake ran fine, BigQuery only worked if either I removed this env var altogether from the configs.env or defined the google app creds like this: `google_application_credentials=/keyfile.json` For this reason I made the change in the Dockerfile (as we ask users to refer to that for creating their confings.env in the first place) However, I encountered a network request error with Databricks when I try to run it through docker (using the same credentials as for python) `_databricks.sql.exc.RequestError: Error during request to server_ `It would be great if someone could double check if it is an issue on their side as well.

Ah yes, looking back I didn't make this clear in the docs - for bigquery, you shouldn't add a google_application_credentials variable to config.env - it is being taken care of by mounting the .json as a volume and docker picks it up with the variable in the dockerfile that points to the location. I did it this way because the keyfile will be a file on the user's local computer that the dockerfile isn't 'aware of' unless you mount the file because it is outside the docker environment.  
<img width="426" alt="Screen Shot 2023-02-22 at 8 18 51 am" src="https://user-images.githubusercontent.com/61371935/220460800-fb23bc44-e064-4c5a-80f5-c49666e02548.png">

It's strange that you had to add an = to get the dockerfile to work, the docs say either way is fine: https://docs.docker.com/engine/reference/builder/#env 

In terms of Databricks, I haven't had time to test the dockerfile (sorry I should have mentioned this), if Ryan is able to debug with you, that would be great.

agnes-kiss committed:
Add versioning for Docker
